service_container/service_locators.rst

```diff
@@ -0,0 +1,155 @@
+        {
+            if (isset($handlerMap[$commandClass = get_class($command)])) {
+                $handler = $handlerMap[$commandClass];
+
+                return $handler->handle($command);
+            }
+        }
+    }
+
+    // ...
+    $commandBus->handle(new FooCommand());
+
+Because only one command is handled at a time, other CommandHandler services are not
+used but unnecessarily instantiated.
+
+Service Locators are intended to solve this problem by giving access to a set of
+identified services while instantiating them only when really needed, providing
+laziness.
+
+Configuration
+-------------
+
+For injecting a service locator in a service, use the `service_locator` argument
+type and define the set of services that needs to be accessible from:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        services:
+
+            app.command_bus:
+                class: AppBundle\CommandBus
+                arguments:
+                    - =service_locator:
+                        AppBundle\FooCommand: '@app.command_handler.foo'
+                        AppBundle\BarCommand: '@app.command_handler.bar'
+
+    .. code-block:: xml
+
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+
+                <service id="app.command_bus" class="AppBundle\CommandBus">
+                    <argument type="service-locator">
+                        <argument key="AppBundle\FooCommand" type="service" id="app.command_handler.foo" />
+                        <argument key="AppBundle\BarCommand" type="service" id="app.command_handler.bar" />
+                    </argument>
+                </service>
+
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        use Symfony\Component\DependencyInjection\Argument\ServiceLocatorArgument;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        //...
+
+        $container
+            ->register('app.command_bus')
+            ->setArguments(array(new ServiceLocatorArgument(array(
+                'AppBundle\FooCommand' => new Reference('app.command_handler.foo'),
+                'AppBundle\BarCommand' => new Reference('app.command_handler.bar'),
+            ))))
+        ;
+
+.. note::
+
+    The services defined in the service locator argument must be keyed,
+    as keys become their unique identifier inside the locator.
+
+.. tip::
+
+    Not only services can be passed accessible through a service locator,
+    but all types which are supported by the configuration format used to
+    configure it.
+
+Usage
+-----
+
+Back to our CommandBus which would now look like::
+
+    // ...
+    use Psr\Container\ContainerInterface;
```

Comment:
Goal is to show the provided PSR-11 implem first, so it's clear what you can do when receiving such argument, including throwable exceptions. Just below there's an example showing that it can be used as a callable and how it can be used